src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingPassBase.cs

```diff
@@ -190,10 +193,177 @@ protected static List<TextChange> AdjustCSharpIndentation(FormattingContext cont
+            foreach (var mapping in csharpDocument.SourceMappings)
+            {
+                var mappingSpan = new TextSpan(mapping.OriginalSpan.AbsoluteIndex, mapping.OriginalSpan.Length);
+                var mappingRange = mappingSpan.AsRange(text);
+                if (!range.LineOverlapsWith(mappingRange))
+                {
+                    // We don't care about this range. It didn't change.
+                    continue;
+                }
+
+                var mappingStartLineIndex = (int)mappingRange.Start.Line;
+                if (context.Indentations[mappingStartLineIndex].StartsInCSharpContext)
+                {
+                    // Doesn't need cleaning up.
+                    // For corner cases like (Range marked with |...|),
+                    // @{
+                    //     if (true} { <div></div>| }|
+                    // }
+                    // We want to leave it alone because tackling it here is really complicated.
+                    continue;
+                }
+
+                if (!ShouldCleanup(context, mappingRange.Start))
+                {
+                    // We don't want to run cleanup on this range.
+                    continue;
+                }
+
+                // @{
+                //     if (true)
+                //     {     
+                //         <div></div>|
+                // 
+                //              |}
+                // }
+                // We want to return the length of the range marked by |...|
+                //
+                var whitespaceLength = text.GetFirstNonWhitespaceOffset(mappingSpan);
+                if (whitespaceLength == null)
+                {
+                    // There was no content here. Skip.
+                    continue;
+                }
+
+                var spanToReplace = new TextSpan(mappingSpan.Start, whitespaceLength.Value);
+                if (!context.TryGetIndentationLevel(spanToReplace.End, out var contentIndentLevel))
+                {
+                    // Can't find the correct indentation for this content. Leave it alone.
+                    continue;
+                }
+
+                // At this point, `contentIndentLevel` should contain the correct indentation level for `}` in the above example.
+                var replacement = context.NewLineString + context.GetIndentationLevelString(contentIndentLevel);
+
+                // After the below change the above example should look like,
+                // @{
+                //     if (true)
+                //     {     
+                //         <div></div>
+                //     }
+                // }
+                var change = new TextChange(spanToReplace, replacement);
+                changes.Add(change);
+            }
+
+            var changedText = text.WithChanges(changes);
+            return changedText;
+        }
+
+        private static bool ShouldCleanup(FormattingContext context, Position position)
```

Comment:
This method is looks pretty similar to `ShouldFormat` method in the `CSharpFormattingPass` but it is slightly different. Leaving it separate for now but will merge them in the future if possible.

Reply:
Is it just the implicit statement difference?

Reply:
For now, yes.